Mathlib/Computability/TuringDegrees.lean

```diff
@@ -0,0 +1,337 @@
+every partial function `g`.
+-/
+theorem partrec_iff_partrec_in_everything
+  (f : ℕ →. ℕ) : Nat.Partrec f ↔ (∀ g, RecursiveIn g f) := by
+  constructor
+  { intro pF
+    intro g
+    induction pF
+    case zero =>
+      apply RecursiveIn.zero
+    case succ =>
+      apply RecursiveIn.succ
+    case left =>
+      apply RecursiveIn.left
+    case right =>
+      apply RecursiveIn.right
+    case pair _ _ _ _ ih1 ih2 =>
+      apply RecursiveIn.pair ih1 ih2
+    case comp _ _ _ _ ih1 ih2 =>
+      apply RecursiveIn.comp ih1 ih2
+    case prec _ _ _ _ ih1 ih2 =>
+      apply RecursiveIn.prec ih1 ih2
+    case rfind _ _ ih =>
+      apply RecursiveIn.rfind ih }
+  { intro H
+    have lem : RecursiveIn (fun _ => pure 0) f := H (fun _ => pure 0)
+    rw [← partrec_iff_partrec_in_zero] at lem
+    exact lem }
+
+/--
+Proof that `turing_reducible` is reflexive.
+-/
+theorem turing_reducible_refl (f : ℕ →. ℕ) : f ≤ᵀ f :=
+  RecursiveIn.oracle
+
+/--
+Proof that `turing_equivalent` is reflexive.
+-/
+theorem turing_equivalent_refl (f : ℕ →. ℕ) : f ≡ᵀ f :=
+  ⟨turing_reducible_refl f, turing_reducible_refl f⟩
+
+/--
+Proof that `turing_equivalent` is symmetric.
+-/
+theorem turing_equivalent_symm {f g : ℕ →. ℕ} (h : f ≡ᵀ g) : g ≡ᵀ f :=
+  ⟨h.2, h.1⟩
+
+/--
+Proof that `turing_reducible` is transitive.
+-/
+theorem turing_reducible_trans {f g h : ℕ →. ℕ} :
+  f ≤ᵀ g → g ≤ᵀ h → f ≤ᵀ h := by
+    intro hg hh
+    induction hg
+    case zero =>
+      apply RecursiveIn.zero
+    case succ =>
+      apply RecursiveIn.succ
+    case left =>
+      apply RecursiveIn.left
+    case right =>
+      apply RecursiveIn.right
+    case oracle =>
+      exact hh
+    case pair f' h' _ _ hf_ih hh_ih =>
+      apply RecursiveIn.pair
+      { apply hf_ih }
+      { apply hh_ih }
+    case comp f' h' _ _ hf_ih hh_ih =>
+      apply RecursiveIn.comp
+      { apply hf_ih }
+      { apply hh_ih }
+    case prec f' h' _ _ hf_ih hh_ih =>
+      apply RecursiveIn.prec
+      { apply hf_ih }
+      { apply hh_ih }
+    case rfind f' _ hf_ih =>
+      apply RecursiveIn.rfind
+      { apply hf_ih }
+
+/--
+Proof that `turing_equivalent` is transitive.
+-/
+theorem turing_equivalent_trans :
+  Transitive turing_equivalent :=
+  fun _ _ _ ⟨fg₁, fg₂⟩ ⟨gh₁, gh₂⟩ =>
+    ⟨turing_reducible_trans fg₁ gh₁, turing_reducible_trans gh₂ fg₂⟩
+
+/--
+Instance declaring that `turing_equivalent` is an equivalence relation.
+-/
+instance : Equivalence turing_equivalent :=
+  {
+    refl := turing_equivalent_refl,
+    symm := turing_equivalent_symm,
+    trans := @turing_equivalent_trans
+  }
+
+/--
+The Turing degrees as the set of equivalence classes under Turing equivalence.
+-/
+def TuringDegree :=
+  Quot turing_equivalent
+
+/--
+The `join` function combines two partial functions `f` and `g` into a single partial function.
+It maps even numbers to the result of `f` applied to half the number,
+and odd numbers to the result of `g` applied to half the number.
```

Comment:
This docstring doesn't look accurate - it doesn't describe the `.map (fun x => 2 * x)` on the result of `f` and the `.map (fun y => 2 * y + 1)` on the result of `g`.